### ISSUE TITLE
SASL priorities changed anannounced

### ISSUE THREAD

**Issue by sualko** [strophe/strophejs#350]
I have some complains (jsxc/jsxc#894 nextcloud/jsxc.nextcloud#149) that the login is not working anymore, because OAUTH is used.

Despite the discussion in #290, it seams that priorities were changed in 846fc25fea1ce476e4a17a839fdaf840461d6b45. Is there any reason for prioritizing OAUTH over SCRAM and why was it not announced? Or did I miss something?

**Comment by jcbrand** [comment]
Sorry, it wasn't my intention to break anything, and I forgot about #290.

I think what happened is that I noticed that the documented order of priorities (see here: https://github.com/strophe/strophejs/commit/846fc25fea1ce476e4a17a839fdaf840461d6b45#diff-7eb52b366866677666470e019283c8eaL3170) didn't match the actual order of priorities in the code, so I updated the code to match the documentation.

Seems like I should have updated the documentation instead.

**Comment by jcbrand** [comment]
Also, given that a specific order of priorities was already documented, I didn't consider it necessary to "announce" something that was already documented.

**Comment by jcbrand** [comment]
That said, it's not clear to me why a client should ever break when SASL priorities are different than what it expects.

It should just ignore the mechanisms it doesn't support or understand.

**Comment by sualko** [comment]
> Seems like I should have updated the documentation instead.

:joy:

> It should just ignore the mechanisms it doesn't support or understand.

Yes, but the client is supporting those mechanism, because Strophe is supporting them. As the documentation stated, the connection object is expecting a password and no token. So the expectation is, at least for me, that a login should work without any additional configuration. At least, we should retry the login if the token was rejected. Maybe it would also be great if we could separate token and password login.

**Comment by jcbrand** [comment]
So this must mean that the XMPP server actually advertises OAuth and then JSXC tries to use that once the user has supplied a password?

BTW, it is possible to tell Strophe what mechanisms you support when instantiating your connection, that way (by omission) you can tell it that OAuth is not supported.

See the "mechanisms" section here: http://strophe.im/strophejs/doc/1.3.5/files/strophe-umd-js.html#Strophe.Connection.Strophe.Connection

I'll set the priorities back to how they were in #290

**Comment by sualko** [comment]
> So this must mean that the XMPP server actually advertises OAuth and then JSXC tries to use that once the user has supplied a password?

That's right, but maybe I should create a separate token login. It was just unneeded until now.

> I'll set the priorities back to how they were in #290

:+1:
